From 92e22ca23e52066da2258df8c76a2dca8a428bcc Mon Sep 17 00:00:00 2001
From: Vasiliy Kulikov <segoon@openwall.com>
Date: Tue, 4 Oct 2011 00:33:15 -0700
Subject: [PATCH] set_interface_var() doesn't check interface name and blindly
 does fopen(path "/" ifname, "w") on it.  As "ifname" is an
 untrusted input, it should be checked for ".." and/or "/"
 in it.  Otherwise, an infected unprivileged daemon may
 overwrite contents of file named "mtu", "hoplimit", etc. in
 arbitrary location with arbitrary 32-bit value in decimal
 representation ("%d").  If an attacker has a local account
 or may create arbitrary symlinks with these names in any
 location (e.g. /tmp), any file may be overwritten with a
 decimal value.

---
 device-linux.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

--- a/device-linux.c
+++ b/device-linux.c
@@ -246,6 +246,10 @@
 	if (!iface[0] || !strcmp(iface, ".") || !strcmp(iface, "..") || strchr(iface, '/'))
 		return -1;
 
+	/* No path traversal */
+	if (strstr(name, "..") || strchr(name, '/'))
+		return -1;
+
 	if (access(spath, F_OK) != 0)
 		return -1;
 
